object/slicer/slicer.go

```diff
@@ -76,82 +82,107 @@ type Slicer struct {
-// written to the metadata of all resulting objects as a creation epoch.
-//
-// See also NewSession.
-func New(signer neofscrypto.Signer, cnr cid.ID, owner user.ID, w ObjectWriter, opts Options) *Slicer {
+func New(ctx context.Context, nw NetworkedClient, signer neofscrypto.Signer, cnr cid.ID, owner user.ID, sessionToken *session.Object) (*Slicer, error) {
+	ni, err := nw.NetworkInfo(ctx, client.PrmNetworkInfo{})
+	if err != nil {
+		return nil, fmt.Errorf("network info: %w", err)
+	}
+
+	opts := Options{
+		objectPayloadLimit: ni.MaxObjectSize(),
+		currentNeoFSEpoch:  ni.CurrentEpoch(),
+		sessionToken:       sessionToken,
+	}
+
+	if !ni.HomomorphicHashingDisabled() {
+		opts.CalculateHomomorphicChecksum()
+	}
+
+	var hdr object.Object
+	hdr.SetContainerID(cnr)
+	hdr.SetType(object.TypeRegular)
+	hdr.SetOwnerID(&owner)
+	hdr.SetCreationEpoch(ni.CurrentEpoch())
+	hdr.SetSessionToken(sessionToken)
+
 	return &Slicer{
-		signer: signer,
-		cnr:    cnr,
-		owner:  owner,
-		w:      w,
 		opts:   opts,
-	}
+		w:      nw,
+		signer: signer,
+		hdr:    hdr,
+	}, nil
 }
 
-// NewSession creates Slicer which generates objects within provided session.
-// NewSession work similar to New with the detail that the session issuer owns
-// the produced objects. Specified session token is written to the metadata of
-// all resulting objects. In this case, the object is considered to be created
-// by a proxy on behalf of the session issuer.
-func NewSession(signer neofscrypto.Signer, cnr cid.ID, token session.Object, w ObjectWriter, opts Options) *Slicer {
-	return &Slicer{
-		signer:       signer,
-		cnr:          cnr,
-		owner:        token.Issuer(),
-		w:            w,
-		opts:         opts,
-		sessionToken: &token,
-	}
+// Put creates new NeoFS object from the input data stream, associates the
+// object with the configured container and writes the object via underlying
+// [ObjectWriter]. After a successful write, Slice returns an [oid.ID] which is a
+// unique reference to the object in the container. Slice sets all required
+// calculated fields like payload length, checksum, etc.
+//
+// Slice allows you to specify [object.Attribute] parameters to be written to the
+// resulting object's metadata. Keys SHOULD NOT start with system-reserved
+// '__NEOFS__' prefix.
+//
+// See [New] for details.
+func (x *Slicer) Put(data io.Reader, attrs []object.Attribute) (oid.ID, error) {
+	x.hdr.SetAttributes(attrs...)
+	return slice(x.w, x.hdr, data, x.signer, x.opts)
+}
+
+// InitPut works similar to [Slicer.Put] but provides [PayloadWriter] allowing
+// the caller to write data himself.
+func (x *Slicer) InitPut(attrs []object.Attribute) (*PayloadWriter, error) {
+	x.hdr.SetAttributes(attrs...)
+	return initPayloadStream(x.w, x.hdr, x.signer, x.opts)
+}
+
+// Put works similar to [Slicer.Put], but allows flexible configuration of object header.
+// The method accepts [Options] for adjusting max object size, epoch, session token, etc.
```

Comment:
It'll duplicate `header` for tokens, epoch and homo.

Reply:
I mean options vs header. Header has a large subset of data from options. Token/epoch for sure. Maybe not homomorphic hash though, it's to be calculated (or not) per slice.

Reply:
Options are generally used for child objects, and we have to have this info for them. Options look like a good place for them.
We may try to extract this info from header object and pass deeper, what do you think about extraction from header?

Reply:
OK, maybe not a big deal and can be handled by SDK user.